### PR TITLE
fixes #22619; don't lift cursor fields in the hook calls

### LINKED ISSUE #22619: Long-term arc-closure iterator problem
### Description

As mentioned over on https://github.com/nim-lang/Nim/issues/22237#issuecomment-1663704779 , I get consistent run-time crashes with this compiled with `nim c --mm:arc -d:broken` (not sure what boolean parity is preferred for workaround defines - feel free to flip the `when` if you like..):

```Nim
import os, posix

proc readFile(path: string, buf: var string, st: ptr Stat=nil, perRead=4096) =
  buf.setLen(0)
  var off = 0
  let fd = open(path, O_RDONLY)
  if fd == -1: return                 #likely vanished between getdents & open
  defer: discard close(fd)
  if st != nil:
    if fstat(fd, st[]) == -1: return  #early return virtually impossible
    if st[].st_size > 0:
      buf.setLen st[].st_size         #may miss actively added; (a race anyway)
      let nRead = read(fd, buf[0].addr, st[].st_size.int)
      if nRead == st[].st_size: return
      if nRead != -1: off = nRead     #fall to loop on a short|failed read
  while true:
    buf.setLen off + perRead          # Ensure room for read()
    let nRead = read(fd, buf[off].addr, perRead)
    if nRead == -1:
      if errno == EAGAIN or errno == EINTR: continue
      return
    elif nRead < perRead:
      buf.setLen(off + nRead)
      break
    off += nRead

proc nonDecimal(s: string): bool =
  for c in s: (if c < '0' or c > '9': return true)

iterator allPids*(): string =
  for pcKind, pid in walkDir("/proc", relative=true):
    if pcKind != pcDir or pid.nonDecimal: continue
    yield pid

when defined(broken):
  proc pidsIt(pids: seq[string]): auto =
    result = iterator(): string =
      if pids.len > 0:
        for pid in pids: yield pid
      else:
        for pid in allPids(): yield pid

  template forPid(pids: seq[string], body) {.dirty.} =
    let it = pidsIt(pids)
    for pid in it(): body
else:
  template forPid(pids: seq[string], body) {.dirty.} =
    if pids.len > 0:
      for pid in pids: body
    else:
      for pid in allPids(): body

var buf = newStringOfCap(4096)          #shared IO buffer for all readFile

proc readStat(pdir: string) = (pdir & "stat").readFile buf

proc readProc(pid: string) = readStat "/proc/" & pid & "/"

proc displayASAP =
  let pids: seq[string] = @[]
  forPid(pids): readProc(pid)
  while true:
    forPid(pids): readProc(pid)

displayASAP()
```

### Nim Version

Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2023-09-01
Copyright (c) 2006-2023 by Andreas Rumpf

### Current Output

```text
$ nim r --mm:arc -d:broken jj
```
gives:
```
Hint: mm: arc; opt: none (DEBUG BUILD, `-d:release` generates faster code)
40620 lines; 0.458s; 69.676MiB peakmem; proj: jj; out: /u/cb/.cache/nim/jj_d/jj_7E4A163C1B678C2C66E32AD4B1AD6F0B5010548A [SuccessX]
Hint: /u/cb/.cache/nim/jj_d/jj_7E4A163C1B678C2C66E32AD4B1AD6F0B5010548A [Exec]
Traceback (most recent call last)
/tmp/jj.nim(65)          jj
/tmp/jj.nim(63)          displayASAP
/usr/lib/nim/lib/std/private/osdirs.nim(239) :anonymous
/usr/lib/nim/lib/std/private/ospaths2.nim(180) /
/usr/lib/nim/lib/std/private/ospaths2.nim(121) joinPath
/usr/lib/nim/lib/std/private/ospaths2.nim(89) joinPathImpl
/usr/lib/nim/lib/pure/pathnorm.nim(107) addNormalizePath
/usr/lib/nim/lib/system/alloc.nim(1073) realloc
/usr/lib/nim/lib/system/alloc.nim(1052) alloc
/usr/lib/nim/lib/system/alloc.nim(868) rawAlloc
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Error: execution of an external program failed: '/u/cb/.cache/nim/jj_d/jj_7E4A163C1B678C2C66E32AD4B1AD6F0B5010548A'
```

### Expected Output

```text
In the example code one expects only an infinite loop making no output.
```

### Possible Solution

Well, as per the example code without `-d:broken` I just gave up on the iterator, but other people may have situations where such workarounds are harder.

Also, `--mm:markAndSweep` or others work fine, but I guess everything but `--mm:arc` is going away "eventually".  So, this is not a long-term workaround.

### Additional Information

AFAICT, this has been true for like 4 years since the beginning of `gc:arc` days, though I have by no means tested every version.

There is probably a smaller example one could construct which might help track down the codegen problem.

EDIT: Also, @ringabout said he reproduced the crash (but not on which nim version, though as mentioned I don't think it's at all sensitive to that).

### PR DESCRIPTION
**Pull request by ringabout**
fixes https://github.com/nim-lang/Nim/issues/22619

It causes double free for closure iterators because cursor fields are destroyed in the lifted destructors of `Env`.

Besides, according to the Nim manual

> In fact, cursor more generally prevents object construction/destruction pairs and so can also be useful in other contexts.

At least, destruction of cursor fields might cause troubles.

todo
- [x] tests
- [x] revert a certain old PR

### CODE REVIEW

File: compiler/liftdestructors.nim [hunk 1]
```diff
@@ -162,8 +162,7 @@ proc fillBodyObj(c: var TLiftCtx; n, body, x, y: PNode; enforceDefaultOp: bool)
     if c.filterDiscriminator != nil: return
     let f = n.sym
     let b = if c.kind == attachedTrace: y else: y.dotField(f)
-    if (sfCursor in f.flags and f.typ.skipTypes(abstractInst).kind in {tyRef, tyProc} and
-        c.g.config.selectedGC in {gcArc, gcAtomicArc, gcOrc, gcHooks}) or
+    if (c.g.config.selectedGC in {gcArc, gcAtomicArc, gcOrc, gcHooks}) or
```
Comment (ringabout) [inline compiler/liftdestructors.nim]:
```suggestion
    if (sfCursor in f.flags and c.g.config.selectedGC in {gcArc, gcAtomicArc, gcOrc, gcHooks}) or
```

**Comment by ringabout** [comment]
https://github.com/actions/checkout/issues/1448 seems to break our CI. I will try to update it to v4

File: tests/iter/t22619.nim
Comment (ringabout) [inline tests/iter/t22619.nim:1]: All credits to @zerbina! Thank you for such detailed tests!

File: compiler/transf.nim [hunk 2]
```diff
@@ -654,7 +654,7 @@ proc findWrongOwners(c: PTransf, n: PNode) =
   else:
     for i in 0..<n.safeLen: findWrongOwners(c, n[i])
 
-proc isSimpleIteratorVar(c: PTransf; iter: PSym; call: PNode; owner: PSym): bool =
+proc isSimpleIteratorVar(c: PTransf; iter: PSym): bool =
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion
proc isSimpleIteratorVar(c: PTransf; iter: PSym; call: PNode; owner: PSym): bool =
```

File: compiler/transf.nim [hunk 3]
```diff
@@ -666,22 +666,9 @@ proc isSimpleIteratorVar(c: PTransf; iter: PSym; call: PNode; owner: PSym): bool
     else:
       for c in n: rec(c, owner, dangerousYields)
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion

  proc recSym(n: PNode; owner: PSym; sameOwner: var bool) =
    case n.kind
    of {nkEmpty..nkNilLit} - {nkSym}: discard
    of nkSym:
      if n.sym.owner != owner:
        sameOwner = false
    else:
      for c in n: recSym(c, owner, sameOwner)
```

File: compiler/transf.nim [hunk 4]
```diff
@@ -666,22 +666,9 @@ proc isSimpleIteratorVar(c: PTransf; iter: PSym; call: PNode; owner: PSym): bool
...
-    else:
-      for c in n: recSym(c, owner, sameOwner)
-
   var dangerousYields = 0
   rec(getBody(c.graph, iter), iter, dangerousYields)
   result = dangerousYields == 0
-  # the parameters should be owned by the owner
-  # bug #22237
-  for i in 1..<call.len:
-    recSym(call[i], owner, result)
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion
  # the parameters should be owned by the owner
  # bug #22237
  for i in 1..<call.len:
    recSym(call[i], owner, result)

```

File: compiler/transf.nim [hunk 5]
```diff
@@ -725,7 +712,7 @@ proc transformFor(c: PTransf, n: PNode): PNode =
       for j in 0..<n[i].len-1:
         addVar(v, copyTree(n[i][j])) # declare new vars
     else:
-      if n[i].kind == nkSym and isSimpleIteratorVar(c, iter, call, n[i].sym.owner):
+      if n[i].kind == nkSym and isSimpleIteratorVar(c, iter):
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion
      if n[i].kind == nkSym and isSimpleIteratorVar(c, iter, call, n[i].sym.owner):
```

File: compiler/transf.nim [hunk 6]
```diff
@@ -674,7 +674,6 @@ proc isSimpleIteratorVar(c: PTransf; iter: PSym; call: PNode; owner: PSym): bool
         sameOwner = false
     else:
       for c in n: recSym(c, owner, sameOwner)
-
   var dangerousYields = 0
```
Comment (ringabout) [inline compiler/transf.nim]:
```suggestion

  var dangerousYields = 0
```

**Comment by Araq** [comment]
Ping @narimiran please backport.

**Pull request merged**